scivision/catalog/catalog.py

```diff
@@ -1,85 +1,221 @@
+    if isinstance(models, CatalogModels):
+        return models
+    elif isinstance(models, (bytes, str, os.PathLike)):
+        models_raw = Path(models).read_text()
+        return CatalogModels.parse_raw(models_raw)
+    elif models is None:
+        models_raw = pkgutil.get_data(__name__, "data/models.json")
+        return CatalogModels.parse_raw(models_raw)
+    else:
+        raise TypeError("Cannot load datasource from unsupported type")
+
+
+class PandasCatalog:
+    def __init__(self, datasources=None, models=None):
         super().__init__()
 
-        datasources_uri = pkgutil.get_data(__name__, "data/datasources.json")
-        models_uri = pkgutil.get_data(__name__, "data/models.json")
+        if isinstance(datasources, pd.DataFrame):
+            self._datasources = datasources
+        else:
+            datasources_cat = _coerce_datasources_catalog(datasources)
+            self._datasources = pd.DataFrame(
+                [ent.dict() for ent in datasources_cat.entries]
+            )
 
-        datasources = pd.read_json(datasources_uri, orient="index")
-        models = pd.read_json(models_uri, orient="index")
+        if isinstance(models, pd.DataFrame):
+            self._models = models
+        else:
+            models_cat = _coerce_models_catalog(models)
+            self._models = pd.DataFrame([ent.dict() for ent in models_cat.entries])
 
-        self._models = models.explode("task").explode("data_format")
-        self._datasources = datasources.explode("task").explode("format")
+    @property
+    def models(self) -> pd.DataFrame:
+        return self._models
 
     @property
-    def _database(self) -> pd.DataFrame:
-        return self._models.merge(
-            self._datasources,
-            suffixes=("_model", "_data"),
-            how="inner",
-            left_on=["task", "data_format"],
-            right_on=["task", "format"],
+    def datasources(self) -> pd.DataFrame:
+        return self._datasources
+
+    def _compatible_models(self, datasource) -> pd.DataFrame:
+        models_compatible_format = self._models[
+            self._models.format == datasource["format"]
+        ]
+
+        models_compatible_format_labels = models_compatible_format[
+            datasource["labels"] | ~models_compatible_format.labels_required
+        ]
+
+        datasource_tasks = pd.DataFrame(datasource["tasks"], columns=["tasks"])
+        models_compatible_tasks = (
+            self._models[["name", "tasks"]]
+            .explode("tasks")
+            .merge(datasource_tasks, on="tasks", suffixes=("_model", "_datasource"),)
+            .name.drop_duplicates()
+        )
+        return models_compatible_format_labels[
+            models_compatible_format_labels.name.isin(models_compatible_tasks)
+        ]
+
+    # Similar to _compatible_models, but for datasources.  Can't
+    # cleanly combine these two functions, due to the asymmetry
+    # between a model's 'labels_required', and datasource 'labels'.
```

Comment:
Can we not standardise and just use 'labels' for both? Since we are updating the schema anyway

Reply:
A model that doesn't require labels can still use a datasource that provides them, but not vice versa, so the checks aren't symmetric - I thought distinct names was also a useful reminder.  I can improve the comment!

Reply:
hmm interesting, maybe even `model_labels` and `dataset_labels` to be extra clear?

Reply:
How about `labels_required` for models (as we have it), but `labels_provided` for datasets?